### DISCUSSION TITLE
useFieldArray: Removing an item causes formState to turn invalid even though it was valid pre-delete and everything is filled in correctly

### DISCUSSION THREAD

**Discussion by Cold-A-Muse** [react-hook-form/react-hook-form#3367] [Q&A]
Dear Community I need some help with a pesky issue that I have been trying to deal with over the last 2 weeks.

**Situation**
1. I initiate a form instance 
-  mode: 'all'
- defaultValues: <some data>
- resolver: YUP schema to validate my Array.
2. On Initial render the formState.isValid = true
3. I add a second entry to my array by using the `append()` function from useFieldArray
4. formState.isValid returns false now since the second entry doesn't match yup schema :+1: 
5. I remove the second entry by using the `remove(index)` function from useFieldArray
**Expectation**
I am expecting the formState.isValid to return true again as the entry which didn't match the schema is removed
**Reality**
The formState.isValid boolean returns true for a short amount before suddenly returning false right after. :-1: 

**Steps to reproduce**
1. Read console initially and verify isValid = true
2. Click the add condition button and verify isValid = false
3. Click the remove button of the second entry and verify isValid returns true before returning *false* right after

Please see https://codesandbox.io/s/white-fire-bknhy?file=/src/TriggerForm.tsx:9771-9784 to try it out yourself.
Thank you so much for you help :heart:

**Comment by bluebill1049** [comment]
Thanks for the detailed issue report, i will take a look at it.

**Reply by bluebill1049** [reply]
could you do some debug on the following CSB?

https://codesandbox.io/s/white-fire-bknhy?file=/src/EditTriggerView.tsx

i have added two more console.log against data and schema validation.

<img width="610" alt="Screen Shot 2020-11-06 at 11 03 20 am" src="https://user-images.githubusercontent.com/10513364/98309876-b2c10000-201f-11eb-871f-2c428c731f93.png">

<img width="576" alt="Screen Shot 2020-11-06 at 10 58 08 am" src="https://user-images.githubusercontent.com/10513364/98309594-f6673a00-201e-11eb-9e52-8d41fd3b9e6b.png">

**Reply by Cold-A-Muse** [reply]
Hey Bill thanks for checking out my issue!
According to the current yup schema:

matchText: yup.string().when("conditionType", {
        is: (conditionType: { label: string; value: string }) => {
          if (conditionType?.value === TriggerConditionType.TEXT_MATCH) {
            console.log("type is MATCH_TEXT");
            return true;
          }
          return false;
        },
        then: yup.string().required(),
        otherwise: yup.string().notRequired()
      }

1. matchText should only be required when conditionType is TEXT_MATCH right? This is not the case as it equals INNER_RANGE instead. This should result in `yup.string().notRequired()` which I assume can accept null values. If it cannot, how come it is accepted on the intial render?

2. How come yupResolver doesn't show this error? Is this a bug or intentional :)? 

Thank you so much again for your help @bluebill1049 :heart:

**Reply by bluebill1049** [reply]
no worries buddy :) I think the first render working, it's probably because validation against the defaultVlaues.

```
resolver: async (data) => {
      console.log(data)
      console.log(await yupResolver(schema)(data))
      return await yupResolver(schema)(data)
    },
```

Do you want to try to do some debugging with the above code? let me know if you find any issue.

**Reply by Cold-A-Muse** [reply]
Hey Bill!
On initial render defaultValues for that field is also null so I'd assume formState should have returned false in that situation as well. I have found that changing `yup.string().notRequired()` to `yup.string().notRequired().nullable()` does indeed fix my error so that's great news :tada: . However, that doesn't explain why it was accepted on intiail render nor does it explain why the error wasn't shown right? i'd should have seen this error in your first comment using the normal way of running the resolver (e.g. `resolver: yupResolver(schema)` correct?

**Reply by bluebill1049** [reply]
yes, that's strange, you want to change `resolver` with my above code snippet and see what's going on? you can see the data as payload and validation result. 👍

**Reply by Cold-A-Muse** [reply]
Yea so I changed it with your snippet and on initial render there is no errors coming out of the yupResolver :flushed: ! 
When I add a new entry there will be two items in the errors object: undefined, and the correct error. Once I delete the second entry suddenly the error in regards to matchText appears on the first one. Here is the CodeSandbox https://codesandbox.io/s/inspiring-poincare-knimy?file=/src/EditTriggerView.tsx

**Reply by bluebill1049** [reply]
what's the difference between the `data` object? it's got to be something different there.

**Reply by Cold-A-Muse** [reply]
Data object on intial render vs when it starts throwing errors you mean?

**Reply by bluebill1049** [reply]
yes 👍

**Reply by Cold-A-Muse** [reply]
Data on intial render:
![image](https://user-images.githubusercontent.com/5823083/98356069-55c74900-2023-11eb-93ea-ca4abcb90022.png)

Data after deleting the faulty entry:
![image](https://user-images.githubusercontent.com/5823083/98356163-77c0cb80-2023-11eb-8a26-1b53197a808a.png)

So it switches to the field from useFieldArray whereas on intial render it just loads in the data directly from the defaultValues property... :confused:

**Reply by bluebill1049** [reply]
that's right, so the `defaultValues` pass the validation schema that you got, not 
<img width="496" alt="Screen Shot 2020-11-06 at 9 35 43 pm" src="https://user-images.githubusercontent.com/10513364/98356654-06f7ce80-2078-11eb-9f20-6362a9eb85fe.png">

**Reply by bluebill1049** [reply]
and `matchText` should be `null` right?

**Reply by Cold-A-Muse** [reply]
Yes exactly! So is this intended behaviour? as in first u load defaultValues and then use the field of useFieldArray? And if so, it should still show me the error when just initiating useForm with `resolver: yupResolver(schema)` right? Because it didn't do that it just gave me isValid = false with an empty form.errors

**Reply by bluebill1049** [reply]
initial render doesn't even have `matchText` tho

**Reply by Cold-A-Muse** [reply]
matchText should be null here yes! you'd expect that notRequired() in yup would cover null values but i guess not :sweat:

**Reply by Cold-A-Muse** [reply]
> initial render doesn't even have `matchText` tho

yea so in initial render it is just undefined so i am not entirely sure where the null is coming from :sweat_smile:

**Reply by bluebill1049** [reply]
do you have an input named `matchText` or set that value manually?

**Reply by Cold-A-Muse** [reply]
So there is a conditionally rendered input where I'd set matchText but only if the conditionType dropdown is set to 'MATCH_TEXT' which is not the case

**Comment by Cold-A-Muse** [comment]
I think what I am more interested in at the moment is why the error was not shown even though there was one. Was this because the field was not currently rendered? :thinking:
